### PR TITLE
add blueprint to backup and restore etcd using kopia 

### PR DESCRIPTION
**Pull request by kale-amruta**
## Change Overview

Add blueprint to backup and restore etcd using kopia

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+actions:
+  backup:
+    outputArtifacts:
+      etcdBackup:
+        kopiaSnapshot: "{{ .Phases.uploadSnapshot.Output.kopiaOutput }}"
+    phases:
+    - func: KubeTask
+      name: takeSnapshot
+      args:
+        image: ghcr.io/kanisterio/kanister-kubectl-1.18:0.80.0
```
Comment (PrasadG193) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:14]:
```suggestion
        image: ghcr.io/kanisterio/kanister-kubectl-1.18:0.81.0
```

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            # exec the snapshot save command
+            kubectl exec -it -n $etcdns $ETCD_POD -c etcd -- sh -c "ETCDCTL_ENDPOINTS=$endpoints etcdctl snapshot save /tmp/etcd-backup.db"
+            # this pod name will be used to copy and remove the snapshot
+            kando output etcdPod $ETCD_POD
+            kando output etcdNS $etcdns
+
+    - func: KubeTask
+      name: uploadSnapshot
+      args:
+        image: ghcr.io/kanisterio/kanister-kubectl-1.18:0.80.0
```
Comment (PrasadG193) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:37]:
```suggestion
        image: ghcr.io/kanisterio/kanister-kubectl-1.18:0.81.0
```

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+          - errexit
+          - -o
+          - pipefail
+          - -c
+          - |
+            kubectl exec -it -n {{ .Phases.takeSnapshot.Output.etcdNS }} "{{ .Phases.takeSnapshot.Output.etcdPod }}" -c etcd -- sh -c "rm -rf  /tmp/etcd-backup.db"
+  restore:
+    inputArtifactNames:
+    - etcdBackup
+    phases:
```
Comment (PrasadG193) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:66]: [nit] Understanding the restore workflow could be difficult to understand without knowing the context. Should we summarize the workflow in the comments?

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+          - -c
+          - |
+            BACKUP_LOCATION='etcd-backup.db.gz'
+            kubectl cp -c etcd {{ .Phases.takeSnapshot.Output.etcdNS }}/{{ .Phases.takeSnapshot.Output.etcdPod }}:/tmp/etcd-backup.db /tmp/etcd-backup.db
+            gzip -c /tmp/etcd-backup.db  | kando location push --profile '{{ toJson .Profile }}' --path "${BACKUP_LOCATION}" --output-name "kopiaOutput" -
+
+    - func: KubeTask
+      name: removeSnapshot
+      args:
+        image: ghcr.io/kanisterio/kanister-kubectl-1.18:0.80.0
```
Comment (PrasadG193) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:53]:
```suggestion
        image: ghcr.io/kanisterio/kanister-kubectl-1.18:0.81.0
```

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+          - |
+            kubectl exec -it -n {{ .Phases.takeSnapshot.Output.etcdNS }} "{{ .Phases.takeSnapshot.Output.etcdPod }}" -c etcd -- sh -c "rm -rf  /tmp/etcd-backup.db"
+  restore:
+    inputArtifactNames:
+    - etcdBackup
+    phases:
+    - func: PrepareData
+      name: restoreFromObjectStore
+      args:
+        image: "ghcr.io/kanisterio/kanister-tools:0.80.0"
```
Comment (PrasadG193) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:70]:
```suggestion
        image: "ghcr.io/kanisterio/kanister-tools:0.81.0"
```

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+          cp /tmp/etcd-backup.db /mnt/data
+  delete:
+    inputArtifactNames:
+    - etcdBackup
+    phases:
+    - func: KubeTask
+      name: deleteFromObjectStore
+      args:
+        namespace: "{{ .Object.metadata.namespace }}"
+        image: "ghcr.io/kanisterio/kanister-tools:0.80.0"
```
Comment (PrasadG193) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:105]:
```suggestion
        image: "ghcr.io/kanisterio/kanister-tools:0.81.0"
```

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+          - pipefail
+          - -c
+          - |
+            kubectl exec -it -n {{ .Phases.takeSnapshot.Output.etcdNS }} "{{ .Phases.takeSnapshot.Output.etcdPod }}" -c etcd -- sh -c "rm -rf  /tmp/etcd-backup.db"
+  restore:
+    # This phase is not actualy performing restore of the etc backup but is used 
+    # to copy backup data. It spins a pod on a master node having label `etcd-backup`. 
+    # The pod is used to download the backup data from the object store 
+    # and copy it to the /mnt/data location of the pv mapped to pvc `pvc-claim-etcd`.
+    # The pv created mounts to the host path /mnt/data on master node where the cluster-ocp-restore
```
Comment (viveksinghggits) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:68]: `The pv created mounts to the`, should we improve this?

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/pv-etcd-backup.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: pv-volume-etcd
```
Comment (viveksinghggits) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/pv-etcd-backup.yaml:4]: pv already has volume in it, so I don't think we need to have volume separately in the name.

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/pv-etcd-backup.yaml [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  labels:
+    type: local
+spec:
+  storageClassName: default
+  capacity:
+    storage: 10Gi
```
Comment (viveksinghggits) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/pv-etcd-backup.yaml:10]: should we also specify that this needs to be changed based on the size of your etcd snapshot.

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+          - sh
+          - -o
+          - errexit
+          - -o
+          - pipefail
+          - -c
+          - |
+            kubectl exec -it -n {{ .Phases.takeSnapshot.Output.etcdNS }} "{{ .Phases.takeSnapshot.Output.etcdPod }}" -c etcd -- sh -c "rm -rf  /tmp/etcd-backup.db"
+  restore:
+    # This phase is not actualy performing restore of the etc backup but is used
```
Comment (viveksinghggits) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:64]:
```suggestion
    # This phase is not actualy performing restore of the etcd data store but is used
```

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    # to copy backup data. It spins a pod on a master node having label `etcd-restore`.
```
Comment (viveksinghggits) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:65]:
```suggestion
    # to copy backup data to one of the leader nodes. It spins a pod on a master node having label `etcd-restore`.
```

File: examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+          - -c
+          - |
+            kubectl exec -it -n {{ .Phases.takeSnapshot.Output.etcdNS }} "{{ .Phases.takeSnapshot.Output.etcdPod }}" -c etcd -- sh -c "rm -rf  /tmp/etcd-backup.db"
+  restore:
+    # This phase is not actualy performing restore of the etcd data store but is used
+    # to copy backup data to one of the leader nodes. It spins a pod on a master node
+    # having label `etcd-restore`. The pod is used to download the backup file from the
+    # object store and copy it to the /mnt/data location of the PV mapped to PVC `pvc-etcd`.
+    # The PV's mount path is /mnt/data on master node where the cluster-ocp-restore.sh
+    # script would be executed.
```
Comment (viveksinghggits) [inline examples/etcd/etcd-in-cluster/ocp/blueprint-v2/etcd-incluster-ocp-blueprint.yaml:69]:
```suggestion
    # This phase is not actualy performing restore of the etcd data store but is used
    # to copy backup data to one of the leader nodes. It spins a pod on a leader node
    # having label `etcd-restore`. The pod is used to download the backup file from the
    # object store and copy it to the /mnt/data location of the PV mapped to PVC `pvc-etcd`.
    # The PV's mount path is /mnt/data on leader node where the cluster-ocp-restore.sh
    # script would be executed. 
```

**Review by viveksinghggits — APPROVED** [review]

**Review by pavannd1 — APPROVED** [review]
LGTM

**Pull request merged**